tests/unit/s2n_tls13_handshake_test.c

```diff
@@ -344,12 +344,79 @@ int main(int argc, char **argv)
+            /* Verify secrets are wiped */
+            for (size_t i = 0; i < conn->psk_params.psk_list.len; i++) {
+                struct s2n_psk *psk;
+                EXPECT_OK(s2n_array_get(&conn->psk_params.psk_list, i, (void**)&psk));
+                EXPECT_NOT_NULL(psk->identity.data);
+                EXPECT_NULL(psk->secret.data);
+                EXPECT_EQUAL(psk->secret.size, 0);
+                EXPECT_NULL(psk->early_secret.data);
+                EXPECT_EQUAL(psk->early_secret.size, 0);
+            }
+
+            EXPECT_SUCCESS(s2n_connection_free(conn));
+        }
+
+        /* PSKs are wiped when chosen PSK is NOT NULL */
+        {
+            struct s2n_connection *conn;
+            EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_SERVER));
+
+            const struct s2n_ecc_preferences *ecc_preferences = NULL;
+            EXPECT_SUCCESS(s2n_connection_get_ecc_preferences(conn, &ecc_preferences));
+            EXPECT_NOT_NULL(ecc_preferences);
+
+            conn->secure.cipher_suite = &s2n_tls13_aes_128_gcm_sha256;
+            conn->secure.server_ecc_evp_params.negotiated_curve = ecc_preferences->ecc_curves[0];
+            EXPECT_SUCCESS(s2n_ecc_evp_generate_ephemeral_key(&conn->secure.server_ecc_evp_params));
+            conn->secure.client_ecc_evp_params[0].negotiated_curve = ecc_preferences->ecc_curves[0];
+            EXPECT_SUCCESS(s2n_ecc_evp_generate_ephemeral_key(&conn->secure.client_ecc_evp_params[0]));
+
+            const uint8_t psk_data[] = "test psk data";
+            const uint8_t secret_data[] = "test secret data";
+            for (size_t i = 0; i < S2N_TEST_PSK_COUNT; i++) {
+                struct s2n_psk *psk = NULL;
+                EXPECT_OK(s2n_array_pushback(&conn->psk_params.psk_list, (void**) &psk));
+                EXPECT_SUCCESS(s2n_psk_init(psk, S2N_PSK_TYPE_EXTERNAL));
+                EXPECT_SUCCESS(s2n_psk_new_identity(psk, psk_data, sizeof(psk_data)));
+                EXPECT_NOT_EQUAL(psk->identity.size, 0);
+                EXPECT_NOT_EQUAL(psk->identity.data, NULL);
+                EXPECT_SUCCESS(s2n_psk_new_secret(psk, secret_data, sizeof(secret_data)));
+                EXPECT_NOT_EQUAL(psk->secret.size, 0);
+                EXPECT_NOT_EQUAL(psk->secret.data, NULL);
+            }
+
+            /* Set chosen PSK */
+            const uint8_t early_secret_data[] = "test early secret data";
+            struct s2n_psk *chosen_psk = NULL;
+            EXPECT_OK(s2n_array_get(&conn->psk_params.psk_list, 0, (void**) &chosen_psk));
+            EXPECT_NOT_NULL(chosen_psk);
+            /* Derive early secrets for chosen PSK */
+            EXPECT_SUCCESS(s2n_realloc(&chosen_psk->early_secret, sizeof(early_secret_data)));
+            memcpy_check(chosen_psk->secret.data, early_secret_data, sizeof(early_secret_data));
+            conn->psk_params.chosen_psk = chosen_psk;
+            conn->psk_params.chosen_psk_wire_index = 0;
```

Comment:
If you want to match the actual situations possible, sometimes (on the client side) all PSKs will have early secrets. You could set the early secret when you set the identity/secret.

I also wouldn't label this "derive" since it's literally just arbitrary data xD